firebase-firestore/src/test/java/com/google/firebase/firestore/model/ObjectValueBuilderTest.java

```diff
@@ -0,0 +1,228 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.model;
+
+import static com.google.firebase.firestore.Values.map;
+import static com.google.firebase.firestore.Values.valueOf;
+import static com.google.firebase.firestore.testutil.TestUtil.field;
+import static junit.framework.TestCase.assertEquals;
+import static junit.framework.TestCase.assertTrue;
+
+import com.google.firebase.firestore.model.protovalue.ObjectValue;
+import com.google.firebase.firestore.model.value.FieldValue;
+import com.google.firestore.v1.Value;
+import java.util.Collections;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(manifest = Config.NONE)
+public class ObjectValueBuilderTest {
+  private Value fooValue = valueOf("foo");
+  private Value barValue = valueOf("bar");
+  private Value emptyObject = valueOf(Collections.emptyMap());
+
+  @Test
+  public void emptyBuilder() {
+    ObjectValue.Builder builder = ObjectValue.emptyObject().toBuilder();
+    ObjectValue object = builder.build();
+    assertEquals(wrapObject(), object);
+  }
+
+  @Test
+  public void setSingleField() {
+    ObjectValue.Builder builder = ObjectValue.emptyObject().toBuilder();
+    builder.set(field("foo"), fooValue);
+    ObjectValue object = builder.build();
+    assertEquals(wrapObject("foo", fooValue), object);
+  }
+
+  @Test
+  public void setEmptyObject() {
+    ObjectValue.Builder builder = ObjectValue.emptyObject().toBuilder();
+    builder.set(field("foo"), emptyObject);
+    ObjectValue object = builder.build();
+    assertEquals(wrapObject("foo", emptyObject), object);
+  }
+
+  @Test
+  public void setMultipleFields() {
+    ObjectValue.Builder builder = ObjectValue.emptyObject().toBuilder();
+    builder.set(field("foo"), fooValue);
+    builder.set(field("bar"), fooValue);
+    ObjectValue object = builder.build();
+    assertEquals(wrapObject("foo", fooValue, "bar", fooValue), object);
+  }
+
+  @Test
+  public void setSuccessorField() {
+    ObjectValue.Builder builder = ObjectValue.emptyObject().toBuilder();
+    builder.set(field("a"), fooValue);
+    builder.set(field("a0"), fooValue);
```

Comment:
Your current successor implementation would break if you also had a value with a character with an ordinal less than 0 (U+0030), e.g. space (U+0020) or bang (U+0021).